src/undo.h

```diff
@@ -86,21 +86,87 @@ struct ScUndoData
 /** Undo information for a CBlock */
 class CBlockUndo
 {
+    /** Magic number read from the value expressing the size of vtxundo vector.
+     *  It is used for distinguish new version of CBlockUndo instance from old ones.
+     *  The maximum number of tx in a block is roughly MAX_BLOCK_SIZE / MIN_TX_SIZE, which is:
+     *   2M / 61bytes =~ 33K = 0x8012
+     * Therefore the magic number must be a number greater than this limit. */
+    static const uint16_t _marker = 0xfec1;
+
+    /** memory only */
+    bool isNewVersion;
+
 public:
     std::vector<CTxUndo> vtxundo; // for all but the coinbase
     uint256 old_tree_root;
     std::map<uint256, ScUndoData> msc_iaundo; // key=scid, value=amount matured at block height
 
-    ADD_SERIALIZE_METHODS;
-
-    template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(vtxundo);
-        READWRITE(old_tree_root);
-        READWRITE(msc_iaundo);
+    /** create as new */
+    CBlockUndo() : isNewVersion(true) {}
+
+    size_t GetSerializeSize(int nType, int nVersion) const
+    {
+        CSizeComputer s(nType, nVersion);
+        NCONST_PTR(this)->Serialize(s, nType, nVersion);
+        return s.size();
+    }   
+
+    template<typename Stream> void Serialize(Stream& s, int nType, int nVersion) const
+    {
+        if (isNewVersion)
+        {
+            WriteCompactSize(s, _marker);
+            ::Serialize(s, (vtxundo), nType, nVersion);
+            ::Serialize(s, (old_tree_root), nType, nVersion);
+            ::Serialize(s, (msc_iaundo), nType, nVersion);
+        }
+        else
+        {
+            ::Serialize(s, (vtxundo), nType, nVersion);
+            ::Serialize(s, (old_tree_root), nType, nVersion);
+        }
+    }   
+
+    template<typename Stream> void Unserialize(Stream& s, int nType, int nVersion)
+    {
+        // reading from data stream to memory
+        vtxundo.clear();
+        isNewVersion = false;
+
+        unsigned int nSize = ReadCompactSize(s);
+        if (nSize == _marker)
+        {
+            // this is a new version of blockundo
+            ::Unserialize(s, (vtxundo), nType, nVersion);
+            ::Unserialize(s, (old_tree_root), nType, nVersion);
+            ::Unserialize(s, (msc_iaundo), nType, nVersion);
+            isNewVersion = true;
+        }
+        else
+        {
+#if 1
+            s.Rewind(GetSizeOfCompactSize(nSize));
+            ::Unserialize(s, (vtxundo), nType, nVersion);
+#else
+            ::AddEntriesInVector(s, vtxundo, nType, nVersion, nSize);
+#endif
+            ::Unserialize(s, (old_tree_root), nType, nVersion);
```

Comment:
If I get you right, these two branches are both valid implementations. I'd go for the  `::AddEntriesInVector` one, since it avoids specifying extra rewind operations.
The reason I'd like to avoid the rewind is because it leverages on fseek which, as I understand, have different semantics for binary or text files.

Reply:
Ok, you are right